### ISSUE TITLE
C.134 - Ensure all data members have the same access level - Does this deserve to be here?

### ISSUE THREAD

**Issue by khatharr** [isocpp/CppCoreGuidelines#302]
You know, I understand the reasoning behind this, but in my own practice it's just pointless 99% of the time and I end up undoing in those cases because I know very well that no one is ever going to inherit from the class in question. Sometimes a public member is a public member and I don't need to hide that behind annoying accessors/mutators. If trivial Java-esque things like this get put into the standard then I'm just going to disable the whole flagging/warning mechanism, and based on the hubbub I'm hearing from other people reading these right now, I know I'm not the only one who feels that way.

I want to be notified about the tools offered by C++ for making things safer and more efficient, especially by avoiding dangerous legacy stuff. I don't want to know about someone's philosophical preferences, regardless of how awesome and correct they may be.

Not everyone works in a production environment and not all code needs to be designed to rigorous production standards. If the tool is simply a nag then people are going to shut it off. Please use it for things that count.

**Comment by galik** [comment]
The wording on this does not differentiate between *public/private* and, say, *private/protected*. Surely it is acceptable for a class to contain some *private* and some *protected* member variables?

**Comment by khatharr** [comment]
Well, the previous entry (C.133) is "Avoid protected data", so I'm assuming the writer would say no.

**Comment by galik** [comment]
Yes I just read that now.

C.133 says "protected data inherently violates the guidance against putting data in base classes"

If I am reading that correctly it seems the guidance is to only use *pure virtual interface* classes and to avoid creating an *abstract base class* (no data in base classes) with common information for all subclsses.

**Comment by hsutter** [comment]
The guidance is that (by default unless you might be in a rare exception):

 

All public: If you’re writing an aggregate bundle-of-variables without an invariant across those variables (just grouping them in a struct for convenience), then all the variables should be public – no encapsulation is needed or desired. (Also, it’s a common convention to declare these using “struct” instead of “class” – the two keywords have essentially the same meaning in C++ but saying “struct” can be a useful mnemonic that you’re doing a C-style thing.)

 

All protected: Protected data is like public data to derived classes, so would only be appropriate if the base class were an aggregate – but you typically don’t inherit from aggregates, so that’s a contradiction and protected data isn’t appropriate.

 

All private: If you’re writing an type that maintains an invariant, then all the variables should be private – it should be encapsulated. This is the vast majority of classes.

 

Mixed: If a class has data at different access levels, its author probably couldn’t decide whether it was supposed to be encapsulated or not.

 

Again, there might rarely be exceptions, such as base class data members that are for debug instrumentation only and don’t participate in the invariant, but the valid exceptions are far rarer than we see them claimed. More commonly, we see access to a data member is relaxed because it’s “convenient” a.k.a. “a hack”. (I work with the creator of MFC, and he still disagrees, but he’s wrong :) …)

 

Herb

**Comment by khatharr** [comment]
That's just a restatement of what still (after years of hearing it) looks like a pointless philosophical preference. If you flag my classes and the reason that you provide is essentially, "You are confused and this is wrong," then I'm not going to change my code; I'm just going to disable the mechanism.

Again, this isn't the first time I've heard this "rule". I've just never been given anything resembling a persuasive argument about why I should observe it. If part of my interface can be served correctly by a variable instead of a function (or pair of functions -.-) then I use a variable. It's never caused me problems for invariants, warts, wombats, radiation leaks, pirate raids, supernovae, *or even minor inconvenience of any kind*.

The attitude that convenience is "a hack" is highly questionable to say the least. You're asserting a principal, failing to justify it, then saying that cases where it plainly doesn't make sense are "wrong". To the critical thinker *that doesn't look good*.

If you want people to accept a principle then the burden of proof is on you! You're saying that these guidelines are supposed to explain to the user why something should be different. What you've got here right now is nothing like persuasive to me, and given your position on inheritance I'm skeptical about whether you can provide a persuasive explanation on this issue!

**Comment by hsutter** [comment]
Thanks for the disclaimer, understood :)

I did think I said more than that. Let me try again:

> If part of my interface can be served correctly by a variable instead of a function

Right, absolutely. The question is simple: **What code is responsible for maintaining a meaningful/correct value for that variable?**

There are exactly two kinds of data members:
* A: Ones that don’t participate in the object’s invariant. Any combination of values for these members is valid.
* B: Ones that do participate in the object’s invariant. Not every combination of values is meaningful (else there’d be no invariant). Therefore all code that has write access to these variables must know about the invariant, know the semantics, and know (and actively implement and enforce) the rules for keeping the values correct.

Data members in category A should just be `public` (or, if you only want derived classes to see them, `protected`). They don’t need encapsulation. All code in the system might as well see and manipulate them.

Data members in category B should be `private` (or, I suppose, `const`). This is because encapsulation is important. To make them non-private and non-const would mean that the object can't control its own state: An unbounded amount of code beyond the class would need to know about the invariant and participate in maintaining it accurately – if these data members were public, that would be all calling code that uses the object; if they were protected, it would be all the code in current and future derived classes. This leads to brittle and tightly coupled code that quickly becomes a nightmare to maintain. Any code that inadvertently sets the data members to an invalid or unexpected combination of values would corrupt the object and all subsequent uses of the object.

The point was just that most classes are either all A or all B.

Occasionally, sure, there are classes that have both A and B. For example, an encapsulated object may contain something like debug instrumentation that isn’t part of the invariant and so falls into category A -- it isn’t really part of the object’s value or meaningful observable state either. In that case, the A parts should be treated as As (made public, or possibly protected if they should be visible only to derived classes) and the B parts should be treated like B’s (private or const).

Is that clearer? I'm happy to try to improve the statement of the guideline further.

If you still think it's wrong I'd be happy to look at a counterexample; actual examples are always helpful.

**Comment by khatharr** [comment]
Const members are a big one that I was going to bring up. I think it would be good to add something about that in the guideline for things like const UUIDS and such. 

For a trivial non-const example, I sometimes use a strategy that relies on a public "visible" flag on objects so that I can skip their render call if they're not enabled. That flag is usually set externally by scripted events that act on the object, and it's read by the loop that goes over the objects when determining what to draw. If I hide that behind a mutator/accessor then they're going to be trivial inlines that compile away to nothing and just make it syntactically nasty to work with the variable:

void setVisible(bool visible) {this.visible = visible;}
bool getVisible() const {return visible;}

This doesn't touch the entity's invariant(s), so it's not really dangerous. I get that exposing invariant-related state would be silly, but in cases like this sometimes a bool is just a bool. If the guideline specifically targets cases where invariant-related members are public then it would make perfect sense (and probably be completely un-testable by the analyzer).

Honestly, I think this would go away if we had attributes, but I know that's non-trivial.

If you're really set on this one, then what's the plan for the 'flag the class' enforcement procedure? Is this something where I could set an option to just specifically ignore C.134 in programs where I use this kind of technique or do I have to add an annoying annotation to every class that uses it?

**Comment by NicolBolas** [comment]
> This doesn't touch the entity's invariant(s), so it's not really dangerous.

It doesn't touch the entity's invariants *now*. But who's to say that it won't?

Maybe your requirements change, such that entities can be attached to one another. And of course, making a parent entity invisible ought to make the child entity invisible too.

What used to be a simple boolean variable no longer can be implemented as such. Either your setting method has to propagate the new state up the hierarchy (to children that weren't explicitly set to invisible), or your getting method needs to check its parents to see if one of them is invisible.

Going from a public variable to a set of accessors is hard. Going from a simple accessor to a complex accessor series is trivial. This is basic refactoring stuff here; accessors exist to allow you to do this.

> Honestly, I think this would go away if we had attributes, but I know that's non-trivial.

I'll assume you said "properties" instead of "attributes". Given that, then it's clear that your issue is not about *writing* accessors as you claimed; it's about *using* accessors.

You would rather do this:

    ent->visible = true;

Instead of this:

    ent->SetVisible(true);

Well I'm sorry, but I just don't care. The most effective way to allow refactoring is to hide all data by default. And if that means you have to use the latter syntax rather than the former, so be it.

In the absence of properties, you have to pick one: minor syntactic convenience or ease of refactoring. And if you're going to establish general programming guidelines, refactoring has to trump minor syntactic convenience every time.

**Comment by khatharr** [comment]
> It doesn't touch the entity's invariants now. But who's to say that it won't?

??? **Me!** I say!
It's my code, and **I say.**

> (vacuous argument)

No. If I wanted a hierarchy of entities then I'd make a composite type, since the existing entity class isn't designed for that kind of usage. If I just wanted to 'attach' entities then they'd still have distinct visibility because otherwise I'd lose capability for no reason.

> Well I'm sorry, but I just don't care.

Yeah, and I don't really care about your authoritarian views on design. *I don't work for you.* More importantly, I don't want to fight the compiler just because you decided something about design practices that's wholly irrelevant to my work.

> Going from a public variable to a set of accessors is hard. Going from a simple accessor to a complex accessor series is trivial. This is basic refactoring stuff here; accessors exist to allow you to do this.

Number of times I've heard this: approaching infiinty
Number of times I've decided to privatize a previously public variable: probably less than 5
Difficulty doing so: 0
Conclusion: not relevant to me

You're looking at your work environment and *dictating* what I ought to be doing based on your own experiences. Look, when I work for someone else I follow their rules and it makes sense to observe things like that in those cases because eventually someone is going to get 'clever' and decide to screw everything up, but **C++ does not belong solely to large organizations writing giant programs that have to be maintained for long periods of time**. It doesn't even belong solely (or even mostly) to professionals. Frankly if you really want to kill the language off then alienating hobbyists is probably a great next step. 

I've had to listen to people for a while now telling me that C# is the way to go and that C++ is a fading dinosaur. I got excited when I saw Bjarne and Herb talking about GSL and clearing up safety problems with the language, but now I'm looking at all this Java inspired masturbatory design bureaucracy. *Guess what I have to listen to now?*

**Comment by germandiagogomez** [comment]
> For a while now people have been telling me that C# is the way to go and that C++ is a fading dinosaur.

Until they see the performance and have to beg back for "the dinosaur". Reality is what it is.

**Comment by villasv** [comment]
>For a while now people have been telling me that C# is the way to go and that C++ is a fading dinosaur. I got excited when I saw Bjarne and Herb talking about GSL and clearing up safety problems with the language, but now I'm looking at all this Java inspired masturbatory design bureaucracy. Guess what I have to listen to now?

That's the great part about guidelines: you can ignore them. I also think that C# properties are the sweetest syntactic sugar ever and I wish C++ had those.

It doesn't and probably it won't. I also don't like the Java style of a zillion of accessors and mutators.
But I totally understand the reasons and still believe its a sound guideline.

What I would suggest (I'm probably not the first one) is that the Guidelines ensuring tools should be capable of letting the user ignore specific sections or, more fine grained, specific rules, that he disagree with.

**Comment by khatharr** [comment]
I find that to be a reasonable proposal. I'm okay with there being design
guidelines but I don't want to fight the compiler over the ones I don't
care about.

**Comment by khatharr** [comment]
No. It's not. It's a pain in the ass that's making demands that aren't relevant to me. If you find that guideline to be valid then I beg you to make use of it, but I don't so please leave me the hell alone. I'm all for there being guidelines like this, but I don't want to have to pay for the ones I don't want to use.

I'm not here to argue with you about whether or not what I know to be false about my code is true. **I'm asking to not be harassed by my tools.**

**Comment by JaminGrey** [comment]
I'm with Khat here. I signed up to GitHub just to add my voice into this (normally a BitBucket user).

I write fairly clean code at a class-design and function-design level (though my system-wide design leaves alot to be desired), and I make a real attempt to keep my code maintainable. I've been programming in C++ for ten years - far less than you, but I'm not a beginner either.

In the projects I work on are in the tens of thousands of lines of code, as opposed to the millions of lines (a quick check says 100,000 non-whitespace lines, including comments, and 719 classes). So a decently sized individual project, but tiny by corporate standards.

Like Khat, my projects are games and game-development related tools.

Most of my classes are a mix of public and private, with occasional protected use (for member-functions only like C. 133 mentions). I am consistent in my coding style within a project. I get what you are saying about this being your perfect ideal given your mindset and ideas, but other people have different views.
Different projects use different paradigms and patterns, which use different guidelines for variable access.

**Comment by NicolBolas** [comment]
@khatharr

Everyone can claim that "what I know to be false about my code is true". For pretty much any rule. People will claim that there's no point in scoped RAII-based cleanup, since they know better than to return in the middle of their scope. And they are not wrong; it is possible to write code perfectly correctly without RAII-based cleanup.

Why should those coders not be "harassed by my tools" too? Don't they have the right to write fragile and unmaintainable code, just like you want to do? You may claim otherwise, but ultimately your argument is no different from theirs: you both think your code is not fragile and unmaintainable.

The guidelines do not forbid things that are 100% guaranteed to be wrong. The purpose of the guidelines is to help prevent users from doing things that we have good evidence will lead to problems.

The problems created by disobeying C.134 are not fictions created by ivory tower intellectuals. They are not speculative or phantasmal issues. These are real issues that real programmers have really encountered on real projects in real codebases. And therefore, having a rule against it is a good thing.

This rule is not unreasonable (though it does contradict C.131, which says not to write simple accessors). It will neither alienate hobbyiests nor will it kill the language. It is most certainly not the product of "Java inspired masturbatory design bureaucracy" (go read up on the Google style guide if you want to know what *that* looks like).

It's just a rule. It's not a language extension; it's a guideline. If you don't follow it, I promise that the C++ police will not break in and arrest you.

@JaminGrey

Much of what I said to Khatharr applies here. Everyone buys into guidelines when they don't interfere with how they personally code. It's the ones that go against the way you've been coding that you chaff against.

That doesn't mean that those guidelines are wrong. Even for you.

Ultimately, the core guidelines recognizes that different projects have different needs. The prohibition on non-exception styles of error reporting alone means that whole classes of projects can't adopt those rules. The guidelines recognize that, so they accept that some people are going to skip the rules they can't work within.

But C.134 is not like exceptions. If you're in a situation where exceptions are not acceptable, then this is generally for reasons other than "I don't want to use them." It's because you're working in a system where the overhead of an exception being thrown is not an option. Or you're working on a platform where exceptions have to be turned off. Or some other inescapable reason.

What you're talking about is a minor bit of syntactic sugar: not having to type `()` at the end of a member name. Hardly an objective need, is it?

**Comment by khatharr** [comment]
There's no *objective* need to demand that other people conform to externally dictated design practices either.

I agree with C.131 for my work and find it mildly amusing that it's present alongside this. I'm not really interested in a 30 page debate about which one to use though. This is more of an issue about whether or not I'm going to have to write a bunch of suppression statements into my code in order to make the compiler not scream at me every time I compile. If it's fully configurable then that would be fine (especially if there's a way to add things to it for shop rules, etc), but the understanding that I have - and do correct me if I'm wrong here - is that this thing is going to just jump up and start complaining about things like this regardless of whether I agree with them or not.

I accepted and started using RAII when I encountered it because I thought about it and saw that it makes sense. When I was initially presented with the mixed access prohibition (a while before that) I started using that as well, but over time I realized that it wasn't helping anything and just had me writing trivial accessors when I didn't want or need them. I considered the issue, measured it against my work, and made a decision. I don't need outsiders arguing with me about it via the flipping compiler. (We have forums for petty arguments. We don't need to weaponize.) When these kinds of guidelines exist in books, references, and similar resources then they're advice. When they invade my actual work-space then they're obnoxious dictates.

Anyway, when I referred to 'attributes' (yes, I meant 'properties') I was considering that there are shortcuts for trivial properties in some other languages, like {get; set;} or attr_accessor. Again, it wouldn't be trivial to do this in C++ because we handle our interfaces differently, but it struck me as being a potential middle ground.

**Comment by villasv** [comment]
I see the main discussion reduced to:
1. You don't think this guideline should be enforced everywhere
2. You don't want you compiler screaming

Solution:
1. It's not, probably will never be. At least for now, it is not the goal to set these guidelines. Like you don't have to compile with -Wall -Werror -pedantic.
2. Turn off the guideline checker features you don't like. If the tool provide this option or not it is not in the scope of this project.

I sincerely don't understand why do you talk like if we were pursuing you and your style.  No one is ruling your code, and no tools are going to harass you. You will surely be able to keep programming as you always had.

> This is more of an issue about whether or not I'm going to have to write a bunch of suppression statements into my code in order to make the compiler not scream at me every time I compile. If it's fully configurable then that would be fine (especially if there's a way to add things to it for shop rules, etc), but the understanding that I have - and do correct me if I'm wrong here - is that this thing is going to just jump up and start complaining about things like this regardless of whether I agree with them or not.

"This thing" won't jump and start complaining because "this thing" is **just a guideline**. If the compiler will enforce the guidelines, print warnings or ignore them is not in the scope this project. This project **is not dictating compile rules**. You should post those questions for the compile developers when they eventually (might even never happen) get interested in implementing guideline checkers.

**Comment by khatharr** [comment]
My understanding is that these guidelines are intended to be used to introduce static analysis as a part of the build process so that the committee can soft deprecate things (or make other shifts, I imagine) for later removal or alteration.

If that's not the case, and this is just for use with separate tools, then there's less concern. I'd still much prefer those tools to be thoroughly configurable though, because annotation on things like pointer math makes a whole lot of sense.

Incidentally, I know I don't have to compile with -Wall -Werror (or equivalent), but life is a hell of a lot better for everyone when I do. That's certainly not something that I'd like to lose, which is large part of why I'm being fussy.

**Comment by JaminGrey** [comment]
> Everyone buys into guidelines when they don't interfere with how they personally code. It's the ones that go against the way you've been coding that you chaff against.

I frequently read and accept new guidelines - even ones that go against the way I code - when they make sense to me and (when I see that they) carry enough value to make it worth following. Rest assured, I'm not still coding the way I was when I started programming, but over the years have frequently improved my coding style through my own insights, and through the insights of others.

Dismissing opposition as "resistant to change" is a cop-out. Claiming "we know right, and everyone else is just stubborn and out-dated" isn't logically sound.

>That doesn't mean that those guidelines are wrong. Even for you.

It's not that we think any idea we don't like is wrong, it's that we think the value gained from enforcing that idea is less than the value gained from ignoring it. There's *tradeoffs* in many of the rules; don't assume that you are getting extra safety for free, but are trading something else for it (whether convenience, or having to write extra boilerplate, or something else).

Once you recognize that, then it should be an easy step to say, "*Hey, these people are suggesting the tradeoff in this circumstance might not be worth it. Maybe I should take a second look?*"

We're not opposed to improving our code's quality. We're not opposed to learning new coding styles and new ideas.

> [exceptions being used as an anology]
Fwiw, I do use exceptions. You may be making an assumption that, because I make games I don't use exceptions. If so, that assumption is mistaken.

> The guidelines recognize that, so they accept that some people are going to skip the rules they can't work within.

Exactly. If the guidelines are actually treated as *guidelines*, and not *rules*, I don't have a problem. If each individual rule can be disabled via the compiler (whichever compiler may be implementing this as a static analyzer), I have zero complaints.

My complaint is merely: Make sure this isn't (by default) *enforced* as a rule, but merely *suggested* as a guideline, and that that suggestion can be disabled *per rule*, rather than having to throw out the entire baby because the bathwater gets muddied.

This may seem obvious to you, 'Of course they'll be disableable!', but just as you can't assume a variable won't later become an invariant, we can't assume a compiler-creator will design the compiler the way we want to use it without us voicing our views.

> What you're talking about is a minor bit of syntactic sugar: not having to type () at the end of a member name.

It was late when I responded. I mistook what the rule was saying, thinking it meant you shouldn't mix private, protected, and public member *functions* in one class, which is ludicrous.

Since it's merely talking about member-*variables*, then I agree with the idea and already do this in my own code - with three or four exceptions (out of ~700 classes). My mistake for my mis-comprehension.

I still want to ensure each individual guideline is togglable by future tools. You said, 'that's outside the scope of this project'. Ofcourse it is, but if we want to voice our concerns here, where the tool creators will see them, rather than having to voice our concerns for each individual tool *after* they get released. We want the guidelines themselves to mention to tool creators that not every user will need to follow every rule, and so they should be disable-able on a per-rule basis.

Ideally, on a per-rule basis, they'd be able to be marked as errors, warnings, or ignored.

**Comment by khatharr** [comment]
That's pretty much where I'm at. If this is simply a guideline then I'd rather the guidance discuss invariants and inappropriate intimacy rather than just saying, "We've been at this a long time and you're confused."

If this gets implemented as something configurable then I have no complaint, but if it's something where I'll have to annotate my code every time then I use a public member variable in a class then I'm disgruntled.

**Comment by NicolBolas** [comment]
> My understanding is that these guidelines are intended to be used to introduce static analysis as a part of the build process so that the committee can soft deprecate things (or make other shifts, I imagine) for later removal or alteration.

I sincerely hope not. Well, the part about "soft deprecation" at any rate.

The language is the language, and I hope the committee isn't going to just decide "well, X has been forbidden by the guidelines for 5 years now, so let's just remove X." And I don't think there is anything in the guidelines that suggests that this is the idea. And I don't recall anything in Bjarne or Herb's talk that suggest that they're going this way.

I do want to see the guidelines be part of a build process. Even coded into compilers. But they should be opt-in, and they should allow you to pick which rules you do and do not want. And this should be true for every rule. Even the invariant ones.

The guidelines should not be used as a back-door way to eventually remove functionality from the language.

> If this is simply a guideline then I'd rather the guidance discuss invariants and inappropriate intimacy rather than just saying, "We've been at this a long time and you're confused."

And yet "we've been at this a long time," is the entire justification for why we want invariants encoded into types. We've been at this a long time. And when invariants *aren't* part of types or aren't actually checked, they will *inevitably* be violated.

Why are you willing to accept this reasoning for invariants but not for C.134?

What's funny is that some of the "invariant" rules (not necessarily in this guideline) is something that I don't much care for. In Herb's presentation at CppCon (Writing Good C++ by Default), VS's static analyzer seems to misrepresent `vector::push_back` as *always* invalidating iterators. This is of course not even remotely close to reality. Indeed, I have code that relies on not invalidating iterators, via explicitly setting the capacity beforehand.

Not a *lot* of code, mind you. Just the one function.

We need attributes for disabling such specific rules within a region of code, to make it explicit that "yes, I'm doing something that is *normally* dubious, but I'm explicitly making it safe." It's up to the code reviewer to then recognize them and judge whether they're doing it correctly.

**Comment by khatharr** [comment]
Regarding the deprecation, I'm looking for it in the videos, but they're talking about getting rid of things that simply don't work anymore or have been completely superseded, like auto_ptr. (I seem to remember it being Chandler that said it?) I'll update this post when I find it.

> And yet "we've been at this a long time," is the entire justification for why we want invariants encoded into types. We've been at this a long time. And when invariants aren't part of types or aren't actually checked, they will inevitably be violated.

That's a self-contradictory paragraph. though. :octocat: You can take out the self-justifying part and use what's left to begin an informative section that people learn something from, rather than resorting to ad verecundiam, which people will tend to ignore. I don't accept ad verecundiam for anything, let alone as an explanation for invariants. I encapsulate my invariants because I don't want something external to come along and put my objects in invalid states. I understand what "invariant" means and why making guarantees about my types makes them stronger. I don't expect other people to take my word for it, though. I want them to look at the reasoning and make their own decision so that they'll know why rather than what.

About the iterators, pb invalidates iterators in the abstract sense rather than in the logical sense. You understand the underlying behavior and reasoning, so you're working around it, but the standard provides no guarantee that iterators will be usable even if you pre-allocated. That is, you're relying on implementation rather than on the standard.

They said in the talks that there will be annotations for disabling rules. I think the format was [[disable:rule]] or something like that.

**Comment by NicolBolas** [comment]
The whole point of having a guideline is to prohibit bad behavior. That's far from "soft deprecation" of anything.

>You can take out the self-justifying part and use what's left to begin an informative section that people learn something from, rather than resorting to ad verecundiam, which people will tend to ignore.

That *you* may ignore sage advice does not mean that others will. I tend to assume that people who've done something for a while know what they're talking about unless I have good evidence that they don't. And students tend to assume likewise of their teachers.

Also, it should be noted that C.134 has a justification. You may choose not to accept it, but it is there.

>About the iterators, pb invalidates iterators in the abstract sense rather than in the logical sense. You understand the underlying behavior and reasoning, so you're working around it, but the standard provides no guarantee that iterators will be usable even if you pre-allocated. That is, you're relying on implementation rather than on the standard.

The standard begs to differ. From C++14, section 23.3.6.3, p6, on `vector::reserve`:

> Remarks: Reallocation invalidates all the references, pointers, and iterators referring to the elements in the sequence. <b>No reallocation shall take place during insertions that happen after a call to `reserve()` until the time when an insertion would make the size of the vector greater than the value of `capacity()`.</b>

From C++14, section 23.3.6.5, p1, on `vector::push_back` (and other insertion):

> Remarks: Causes reallocation if the new size is greater than the old capacity. <b>If no reallocation happens, all the iterators and references before the insertion point remain valid.</b>

(emphasis added)

So I am relying on entirely standard-defined behavior.

**Comment by hsutter** [comment]
Updated C.134 to reflect my notes above.
